### ISSUE TITLE
Development Workflow

### ISSUE THREAD

**Issue by gdevanla** [bos/aeson#733]
One challenge I have had while making incremental changes to the code base was that running `stack build` or `stack test` takes a while. What is the workflow other contributors follow in these scenarios?  

At the least, when I run `stack test`, I would like to have a way to just run specific tests or just tests within a particular scope. Currently, when I run `stack test`,a lot of time is spent in building libraries like `criterion` and 'statistics`. Is there a way to turn them off as needed.

Overall, I would like to contribute to a document (assuming this is not there already) that describes a development workflow which other contributors could use.  Can someone point me to any resources or share their workflow. I can create a document by compiling those.

**Comment by Lysxia** [comment]
I also wonder about this. `criterion` takes a loooong time to build. However it's only used by benchmarks. Does your `stack.yaml` mention the `benchmarks/` directory? If so, it should be okay to remove it.

If you do want to build benchmarks, there is still a hacky way to avoid rebuilding criterion, which is to rename the version of aeson you're working on so that criterion doesn't depend on it. I don't know whether there is a more principled solution.

+1 to having this documented somewhere.

**Comment by bergmark** [comment]
I usually use ghci while developing. For stack: `stack ghci --stack-yaml stack-nightly.yaml --test` 
which doesn't include benchmarks. `:r` to reload and `main` to run the test suite. Speeds things up a lot!

The tests are run with tasty, I forget how you filter which tests to run, but it seems to be possible e.g. using `--pattern`.

It think CONTRIBUTING.md is a good place for this info.

Thanks in advance

**Comment by bergmark** [comment]
You can apply cmd line args inside ghci using `withArgs args main`

**Comment by phadej** [comment]
@bergmark also

```
:main --pattern Foo
```

**Comment by phadej** [comment]
Also adding `historyDuplicates: IgnoreConsecutive` to `~/.haskeline` would reduce the `:r` s in your history, improves GHCi driven development hugely. https://github.com/judah/haskeline/wiki/UserPreferences

**Comment by gdevanla** [comment]
With the latest changes on master, I am currently seeing this error.  Any tips on how to handle this would be appreciated!

```
$ stack ghci --stack-yaml stack-nightly.yaml --test
Using main module: 1. Package `aeson' component aeson:test:tests with main-is file: /home/devanla/fsf/aeson/tests/Tests.hs

Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for aeson-1.4.5.0(+fast):
    Diff-0.3.4 from stack configuration does not match >=0.4 && <0.5  (latest matching version is 0.4.0)
needed since aeson is a build target.

Some different approaches to resolving this:

  * Set 'allow-newer: true' in /home/devanla/.stack/config.yaml to ignore all version constraints and build anyway.

  * Recommended action: try adding the following to your extra-deps in ~/fsf/aeson/stack-nightly.yaml:

- Diff-0.4.0@sha256:b5cfbeed498f555a18774ffd549bbeff7a24bdfe5984154dcfc9f4328a3c2847,1275


Error: Plan construction failed.

```

**Comment by bergmark** [comment]
Ah sorry, I forgot that i updated it locally. Try this:
```
resolver: nightly-2019-10-14
packages:
- '.'
- attoparsec-iso8601
flags:
  aeson:
    fast: true
  attoparsec-iso8601:
    fast: true
extra-deps:
- generic-deriving-1.12.4
```

**Comment by gdevanla** [comment]
Thanks, that works. Shouldn't this change be on master?

**Comment by bergmark** [comment]
Indeed it should

**Comment by bergmark** [comment]
Resolved by #736
